File: src/main/java/com/mrpowergamerbr/loritta/frontend/views/GlobalHandler.java
package com.mrpowergamerbr.loritta.frontend.views;

import com.mitchellbosecke.pebble.template.PebbleTemplate;
import com.mrpowergamerbr.loritta.Loritta;
import com.mrpowergamerbr.loritta.LorittaLauncher;
import com.mrpowergamerbr.loritta.frontend.LorittaWebsite;
import com.mrpowergamerbr.loritta.frontend.utils.PaniniUtils;
import com.mrpowergamerbr.loritta.frontend.utils.RenderContext;
import com.mrpowergamerbr.loritta.userdata.LorittaProfile;
import com.mrpowergamerbr.temmiediscordauth.TemmieDiscordAuth;
import lombok.experimental.ExtensionMethod;
import org.apache.commons.lang3.ArrayUtils;
import org.jooby.Request;
import org.jooby.Response;

import java.io.PrintWriter;
import java.io.StringWriter;
import java.lang.management.ManagementFactory;
import java.util.HashMap;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.TimeUnit;

/**
 * Based on PaniniCMS's GlobalHandler: https://github.com/PaniniCMS/Panini/blob/master/src/com/paninicms/views/GlobalHandler.java
 */
@ExtensionMethod(PaniniUtils.class)
public class GlobalHandler {
    public static void render(Request req, Response res) {
        try { // The super try block(tm)
            String[] arguments = req.path().split("/");
            if (arguments.length > 0) {
                arguments = ArrayUtils.remove(arguments, 0); // O primeiro sempre será vazio, então vamos remover ele
            }

            Map<String, Object> contextVars = new HashMap<String, Object>(); // Variáveis usadas no Pebble

            contextVars.put("websiteUrl", LorittaWebsite.getWebsiteUrl());
            contextVars.put("clientId", Loritta.getClientId());
            contextVars.put("websiteUrl", LorittaWebsite.getWebsiteUrl());
            contextVars.put("totalServers", LorittaLauncher.getInstance().getLorittaShards().getGuilds().size());
            contextVars.put("totalUsers", LorittaLauncher.getInstance().getLorittaShards().getUsers().size());
            contextVars.put("epochMillis", System.currentTimeMillis());
            Object temmieObj = null;
            if (req.session().isSet("discordAuthValue")) {
                temmieObj = LorittaWebsite.getOAuth2()
                        .getOrDefault(req.session().get("discordAuthCode").value(), null);
            }
            contextVars.put("temmie", temmieObj);

            long jvmUpTime = ManagementFactory.getRuntimeMXBean().getUptime();

            long days = TimeUnit.MILLISECONDS.toDays(jvmUpTime);
            jvmUpTime -= TimeUnit.DAYS.toMillis(days);
            long hours = TimeUnit.MILLISECONDS.toHours(jvmUpTime);
            jvmUpTime -= TimeUnit.HOURS.toMillis(hours);
            long minutes = TimeUnit.MILLISECONDS.toMinutes(jvmUpTime);
            jvmUpTime -= TimeUnit.MINUTES.toMillis(minutes);
            long seconds = TimeUnit.MILLISECONDS.toSeconds(jvmUpTime);

            StringBuilder sb = new StringBuilder(64);
            sb.append(days);
            sb.append(" dias");
            sb.append(", " + hours);
            sb.append(" horas");
            sb.append(", " + minutes);
            sb.append(" minutos");
            sb.append(" e " + seconds);
            sb.append(" segundos");
            contextVars.put("uptime", sb.toString());

            RenderContext context = new RenderContext()
                    .contextVars(contextVars)
                    .request(req)
                    .response(res)
                    .arguments(arguments);

            Object obj = null;
            if (arguments.length == 0) {
                obj = HomeView.render(context);
            } else if (arguments.is(0, "cmdvis")) {
                obj = CommandVisualizerRender.render(context);
            } else if (arguments.is(0, "doar")) {
                obj = DonateView.render(context);
            } else if (arguments.is(0, "comandos")) {
                obj = CommandsView.render(context);
            } else if (arguments.is(0, "servers")) {
                obj = ServerListView.render(context);
            } else if (arguments.is(0, "auth")) {
                if (req.param("code").isSet()) { // Se o code está marcado, então é um pedido de autenticação via OAuth2
                    String code = req.param("code").value();

                    TemmieDiscordAuth temmie = new TemmieDiscordAuth(code, "https://loritta.website/auth", Loritta.getClientId(), Loritta.getClientSecret(), true);

                    temmie.doTokenExchange(); // Faça o token exchange
                    System.out.println("Token exchanged!");
                    UUID uuid = UUID.randomUUID();
                    LorittaWebsite.getOAuth2().put(uuid.toString(), temmie); // TODO: Tenho certeza que existe um jeito melhor para isto. :/
                    req.session().set("discordAuthCode", uuid.toString()); // Vamos guardar o auth token
                    System.out.println("OAuth2!");
                    try {
                        res.redirect("https://loritta.website/config/servidores");
                    } catch (Throwable ex) {
                        ex.printStackTrace();
                    }
                    return;
                } else if (req.session().get("discordAuthCode").isSet()) { // Se discordAuthCode existe, então já existe uma conta OAuth2 ativada!
                    System.out.println("Trying to reuse Authentication Code...");
                    if (temmieObj == null) { // Se é null, quer dizer que o TemmieDiscordAuth expirou
                        System.out.println("null!");
                        try {
                            res.redirect("https://discordapp.com/oauth2/authorize?redirect_uri=https://loritta.website%2Fauth&scope=identify%20guilds&response_type=code&client_id=297153970613387264");
                        } catch (Throwable ex) {
                            ex.printStackTrace();
                        }
                        return;
                    } else {
                        System.out.println("not null!");
                        TemmieDiscordAuth temmie = (TemmieDiscordAuth) temmieObj;

                        if (false /* !temmie.isValid() */) {
                            System.out.println("but it isn't valid!");
                            // Vamos tentar pegar um novo access token
                            try {
                                System.out.println("doing token exchange!");
                                temmie.doTokenExchangeUsingRefreshToken();
                            } catch (Exception e) {
                                e.printStackTrace();
                                // Inválido, redirecione para o oauth2 request
                                try {
                                    res.redirect("https://discordapp.com/oauth2/authorize?redirect_uri=https://loritta.website%2Fauth&scope=identify%20guilds&response_type=code&client_id=297153970613387264");
                                } catch (Throwable ex) {
                                    ex.printStackTrace();
                                }
                                return;
                            }
                        } else {
                            // Código OAuth2 é válido, vamos redirecionar para o gerenciador de servidores
                            try {
                                System.out.println("OAuth2 valid!");
                                res.redirect("https://loritta.website/config/servidores");
                            } catch (Throwable ex) {
                                ex.printStackTrace();
                            }
                            return;
                        }
                    }
                }
                try {
                    res.redirect("https://discordapp.com/oauth2/authorize?redirect_uri=https://loritta.website%2Fauth&scope=identify%20guilds&response_type=code&client_id=297153970613387264");
                } catch (Throwable ex) {
                    ex.printStackTrace();
                }
                return;
            } else if (arguments.is(0, "config")) {
                // Behind Authentication
                try {
                    TemmieDiscordAuth temmie = (TemmieDiscordAuth) LorittaWebsite.getOAuth2().get(req.session().get("discordAuthCode").value());
                    context.contextVars().put("temmie", temmie);
                    LorittaProfile profile = LorittaLauncher.getInstance().getLorittaProfileForUser(temmie.getCurrentUserIdentification().getId());
                    context.contextVars().put("userProfile", profile);

                    if (!profile.isBanned()) { // Se o usuário não está banido...
                        // code
                        // guild_id
                        // permissions
                        if (context.request().path().equalsIgnoreCase("/config/servidores")) {
                            if (req.param("guild_id").isSet()) {
                                // Se está marcado, então a Loritta foi adicionada em algum servidor
                                try {
                                    res.redirect("https://loritta.website/config/servidor/" + req.param("guild_id").value());
                                } catch (Throwable e) {
                                    e.printStackTrace();
                                }
                                return;
                            }
                            obj = ChooseServerView.render(context, temmie);
                        }
                        if (context.request().path().startsWith("/config/servidor")) {
                            String[] split = context.request().path().split("/");
                            if (split.length >= 4) {
                                String guildId = split[3];
                                System.out.println("reading guildId..." + guildId);
                                if (LorittaLauncher.getInstance().getLorittaShards().getGuildById(guildId) != null) {
                                    // oh shit
                                    obj = ConfigureServerView.render(context, temmie, guildId);
                                } else {
                                    try {
                                        res.redirect("https://discordapp.com/oauth2/authorize?client_id=" + Loritta.getClientId() + "&scope=bot&guild_id=" + guildId + "&response_type=code&redirect_uri=https://loritta.website/config/servidores&permissions=2097176631");
                                    } catch (Throwable e) {
                                        e.printStackTrace();
                                    }
                                    return;
                                }
                            }
                        }
                    } else { // Mas se ele estiver, carregue a página de bans
                        obj = BannedView.render(context, temmie);
                    }
                } catch (Exception e) {
                    e.printStackTrace();
                    obj = "<meta http-equiv=\"refresh\" content=\"0; url=https://discordapp.com/oauth2/authorize?redirect_uri=https://loritta.website%2Fauth&scope=identify%20guilds&response_type=code&client_id=297153970613387264\" />";
                }
            }
            String output = null;

            if (obj instanceof PebbleTemplate) {
                PebbleTemplate compiledTemplate = (PebbleTemplate) obj;
                StringWriter writer = new StringWriter();
                compiledTemplate.evaluate(writer, context.contextVars());
                output = writer.toString();
            } else if (obj instanceof String) {
                output = (String) obj;
            }

            try {
                if (output == null) {
                    output = "Missing output";
                }
                res.send(output);
            } catch (Throwable e) {
                e.printStackTrace();
            }
        } catch (Exception e) {
            StringWriter db1 = new StringWriter();
            PrintWriter coll1 = new PrintWriter(db1);
            e.printStackTrace(coll1);
            try {
                res.send(db1.toString());
            } catch (Throwable ex) {
                e.printStackTrace();
            }
        }
    }
}

File: src/main/java/com/mrpowergamerbr/loritta/frontend/views/GlobalHandler.kt
package com.mrpowergamerbr.loritta.frontend.views

import com.github.salomonbrys.kotson.fromJson
import com.google.common.collect.Lists
import com.mrpowergamerbr.loritta.Loritta
import com.mrpowergamerbr.loritta.Loritta.Companion.GSON
import com.mrpowergamerbr.loritta.LorittaLauncher
import com.mrpowergamerbr.loritta.commands.AbstractCommand
import com.mrpowergamerbr.loritta.frontend.LorittaWebsite
import com.mrpowergamerbr.loritta.frontend.evaluate
import com.mrpowergamerbr.loritta.frontend.views.subviews.*
import com.mrpowergamerbr.loritta.frontend.views.subviews.api.*
import com.mrpowergamerbr.loritta.frontend.views.subviews.api.config.APIGetApiKeyView
import com.mrpowergamerbr.loritta.frontend.views.subviews.api.config.APIGetServerConfigView
import com.mrpowergamerbr.loritta.frontend.views.subviews.api.config.APIUpdateServerConfigView
import com.mrpowergamerbr.loritta.frontend.views.subviews.api.serverlist.*
import com.mrpowergamerbr.loritta.frontend.views.subviews.configure.*
import com.mrpowergamerbr.loritta.utils.LorittaUtilsKotlin
import com.mrpowergamerbr.loritta.utils.loritta
import com.mrpowergamerbr.loritta.utils.oauth2.TemmieDiscordAuth
import org.apache.commons.lang3.exception.ExceptionUtils
import org.jooby.Request
import org.jooby.Response
import org.slf4j.LoggerFactory
import java.lang.management.ManagementFactory
import java.text.MessageFormat
import java.util.*
import java.util.concurrent.TimeUnit

object GlobalHandler {
	var views = mutableListOf<AbstractView>()
	var apiViews = mutableListOf<NoVarsView>()

	val logger = LoggerFactory.getLogger(AbstractCommand::class.java)

	fun render(req: Request, res: Response): String {
		logger.info(" ${req.header("X-Forwarded-For").value()}: ${req.path()}")

		if (req.path().matches(Regex("^/dashboard/configure/[0-9]+/testmessage")) || req.path().matches(Regex("^\\/dashboard\\/configure\\/[0-9]+(\\/)(save)"))) {
			val last = loritta.apiCooldown.getOrDefault(req.header("X-Forwarded-For").value(), 0L)

			val diff = System.currentTimeMillis() - last
			if (2500 >= diff) {
				return GSON.toJson(mapOf("api:code" to LoriWebCodes.RATE_LIMITED, "api:message" to "RATE_LIMITED"))
			}

			loritta.apiCooldown[req.header("X-Forwarded-For").value()] = System.currentTimeMillis()
		}

		apiViews.filter { it.handleRender(req, res, req.path()) }
				.forEach { return it.render(req, res, req.path()) }

		val variables = mutableMapOf<String, Any?>("discordAuth" to null)
		variables["epochMillis"] = System.currentTimeMillis()

		// TODO: Deprecated
		val acceptLanguage = req.header("Accept-Language").value("en-US")

		// Vamos parsear!
		val ranges = Lists.reverse<Locale.LanguageRange>(Locale.LanguageRange.parse(acceptLanguage))

		val defaultLocale = LorittaLauncher.loritta.getLocaleById("default")
		var lorittaLocale = LorittaLauncher.loritta.getLocaleById("default")

		var localeId: String? = null

		for (range in ranges) {
			localeId = range.range.toLowerCase()
			var bypassCheck = false
			if (localeId == "pt-br" || localeId == "pt") {
				localeId = "default"
				bypassCheck = true
			}
			if (localeId == "en") {
				localeId = "en-us"
			}
			val parsedLocale = LorittaLauncher.loritta.getLocaleById(localeId)
			if (bypassCheck || defaultLocale !== parsedLocale) {
				lorittaLocale = parsedLocale
			}
		}

		if (req.param("force_locale").isSet) {
			req.session()["forceLocale"] = req.param("force_locale").value()
		}

		if (req.param("logout").isSet) {
			req.session().destroy()
		}

		if (req.session().isSet("forceLocale")) {
			lorittaLocale  = LorittaLauncher.loritta.getLocaleById(req.session()["forceLocale"].value())
		}

		if (req.param("locale").isSet) {
			lorittaLocale = LorittaLauncher.loritta.getLocaleById(req.param("locale").value())
		}

		// Para deixar tudo organizadinho (o Google não gosta de locales que usem query strings ou cookies), nós iremos usar subdomínios!
		val languageCode = req.path().split("/").getOrNull(1)

		if (languageCode != null) {
			lorittaLocale = when (languageCode) {
				"br" -> LorittaLauncher.loritta.getLocaleById("default")
				"pt" -> LorittaLauncher.loritta.getLocaleById("pt-pt")
				"us" -> LorittaLauncher.loritta.getLocaleById("en-us")
				"es" -> LorittaLauncher.loritta.getLocaleById("es-es")
				else -> lorittaLocale
			}
		}

		for (locale in lorittaLocale.strings) {
			variables[locale.key] = MessageFormat.format(locale.value)
		}

		variables["guildCount"] = loritta.guildCount
		variables["userCount"] = loritta.userCount
		variables["availableCommandsCount"] = loritta.commandManager.commandMap.size
		variables["commandMap"] = loritta.commandManager.commandMap
		variables["executedCommandsCount"] = LorittaUtilsKotlin.executedCommands
		variables["path"] = req.path()
		var pathNoLanguageCode = req.path()
		val split = pathNoLanguageCode.split("/").toMutableList()
		val languageCode2 = split.getOrNull(1)

		val hasLangCode = languageCode2 == "br" || languageCode2 == "es" || languageCode2 == "us" || languageCode2 == "pt"
		if (hasLangCode) {
			split.removeAt(0)
			split.removeAt(0)
			pathNoLanguageCode = "/" + split.joinToString("/")
		} else {
			// Nós iremos redirecionar o usuário para a versão correta para ele, caso esteja acessando o "website errado"
			if (localeId != null) {
				if (req.path() != "/dashboard" && req.path() != "/auth" && !req.path().matches(Regex("^\\/dashboard\\/configure\\/[0-9]+(\\/)(save)")) && !req.path().matches(Regex("^/dashboard/configure/[0-9]+/testmessage")) && !req.path().startsWith("/translation") /* DEPRECATED API */) {
					res.status(302) // temporary redirect / no page rank penalty (?)
					if (localeId == "default") {
						res.redirect("https://loritta.website/br${req.path()}")
					}
					if (localeId == "en-us") {
						res.redirect("https://loritta.website/us${req.path()}")
					}
				}
			}
		}

		variables["pathNL"] = pathNoLanguageCode // path no language code
		variables["loriUrl"] = LorittaWebsite.WEBSITE_URL + "${languageCode2 ?: "us"}/"

		val isPatreon = mutableMapOf<String, Boolean>()
		val isDonator = mutableMapOf<String, Boolean>()

		val lorittaGuild = com.mrpowergamerbr.loritta.utils.lorittaShards.getGuildById("297732013006389252")

		if (lorittaGuild != null) {
			val rolePatreons = lorittaGuild.getRoleById("364201981016801281") // Pagadores de Aluguel
			val roleDonators = lorittaGuild.getRoleById("334711262262853642") // Doadores

			val patreons = lorittaGuild.getMembersWithRoles(rolePatreons)
			val donators = lorittaGuild.getMembersWithRoles(roleDonators)

			patreons.forEach {
				isPatreon[it.user.id] = true
			}
			donators.forEach {
				isDonator[it.user.id] = true
			}
		}

		variables["clientId"] = Loritta.config.clientId
		variables["isPatreon"] = isPatreon
		variables["isDonator"] = isDonator
		var jvmUpTime = ManagementFactory.getRuntimeMXBean().uptime

		val days = TimeUnit.MILLISECONDS.toDays(jvmUpTime)
		jvmUpTime -= TimeUnit.DAYS.toMillis(days)
		val hours = TimeUnit.MILLISECONDS.toHours(jvmUpTime)
		jvmUpTime -= TimeUnit.HOURS.toMillis(hours)
		val minutes = TimeUnit.MILLISECONDS.toMinutes(jvmUpTime)
		jvmUpTime -= TimeUnit.MINUTES.toMillis(minutes)
		val seconds = TimeUnit.MILLISECONDS.toSeconds(jvmUpTime)

		val correctUrl = LorittaWebsite.WEBSITE_URL.replace("https://", "https://$languageCode.")
		variables["uptimeDays"] = days
		variables["uptimeHours"] = hours
		variables["uptimeMinutes"] = minutes
		variables["uptimeSeconds"] = seconds
		variables["currentUrl"] = correctUrl + req.path().substring(1)
		variables["localeAsJson"] = GSON.toJson(lorittaLocale.strings)
		variables["websiteUrl"] = LorittaWebsite.WEBSITE_URL

		if (req.session().isSet("discordAuth")) {
			val discordAuth = Loritta.GSON.fromJson<TemmieDiscordAuth>(req.session()["discordAuth"].value())
			try {
				discordAuth.isReady(true)
				val userIdentification = discordAuth.getUserIdentification() // Vamos pegar qualquer coisa para ver se não irá dar erro
				variables["discordAuth"] = discordAuth
				variables["userIdentification"] = userIdentification
			} catch (e: Exception) {
				req.session().unset("discordAuth")
			}
		}

		try {
			views.filter { it.handleRender(req, res, pathNoLanguageCode, variables) }
					.forEach { return it.render(req, res, pathNoLanguageCode, variables) }
		} catch (e: Exception) {
			logger.error("Erro ao processar conteúdo para ${req.header("X-Forwarded-For").value()}: ${req.path()}", e)
			throw e
		}

		res.status(404)
		return evaluate("404.html", variables)
	}

	fun generateViews()  {
		val views = mutableListOf<AbstractView>()
		val apiViews = mutableListOf<NoVarsView>()

		// ===[ APIS ]===
		apiViews.add(APIGetCommunityInfoView())
		apiViews.add(APIGetChannelInfoView())
		apiViews.add(APIGetRssFeedTitleView())
		apiViews.add(APIGetTwitchInfoView())
		apiViews.add(APILoriGetBalanceView())
		apiViews.add(APILoriGiveBalanceView())
		apiViews.add(APILoriSetBalanceView())
		apiViews.add(APILoriWithdrawBalanceView())
		apiViews.add(APIGetServerSampleView())
		apiViews.add(APIGetServerInformationView())
		apiViews.add(APIVoteServerView())
		apiViews.add(APIGetServerConfigView())
		apiViews.add(APIUpdateServerConfigView())
		apiViews.add(APIGetServersView())
		apiViews.add(APIGetLocaleView())
		apiViews.add(APILoriDailyRewardView())
		apiViews.add(APILoriDailyRewardStatusView())
		apiViews.add(APIJoinServerView())
		apiViews.add(APIGetApiKeyView())
		apiViews.add(APIGetServerVotesView())
		apiViews.add(APIGetCommandsView())

		views.add(HomeView())
		views.add(TranslationView())
		views.add(DashboardView())
		views.add(LorigotchiView())
		views.add(LoriPartnerView())
		views.add(ServersView())
		views.add(FanArtsView())
		views.add(DonateView())
		views.add(CommandsView())
		views.add(DailyView())
		views.add(SupportView())
		// views.add(NashornDocsView())
		views.add(PatreonCallbackView())
		views.add(AuthPathRedirectView())

		views.add(ConfigureServerView())
		views.add(ConfigureEventLogView())
		views.add(ConfigureInviteBlockerView())
		views.add(ConfigureAutoroleView())
		views.add(ConfigurePermissionsView())
		views.add(ConfigureWelcomerView())
		views.add(ConfigureStarboardView())
		views.add(ConfigureAminoView())
		views.add(ConfigureYouTubeView())
		views.add(ConfigureLivestreamView())
		views.add(ConfigureRSSFeedsView())
		views.add(ConfigureNashornCommandsView())
		views.add(ConfigureMusicView())
		views.add(ConfigureEventHandlersView())
		views.add(ConfigureCommandsView())
		views.add(ConfigureTextChannelsView())
		views.add(ConfigureModerationView())
		views.add(ConfigureLoriPartnerView())
		views.add(TestMessageView())

		this.views = views
		this.apiViews = apiViews
	}
}